src/Ubidots.cpp

```diff
@@ -0,0 +1,227 @@
+
+bool Ubidots::send(const char *device_label) {
+  return send(device_label, device_label);
+}
+
+bool Ubidots::send(const char *device_label, const char *device_name) {
+  return _cloudProtocol->send(device_label, device_name);
+}
+
+float Ubidots::get(const char *device_label, const char *variable_label) {
+  if (_debug) {
+    Serial.println("Constructing get Request!");
+  }
+
+  _cloudProtocol->get(device_label, variable_label);
+}
+
+void Ubidots::setDebug(bool debug) {
+  _debug = debug;
+  _cloudProtocol->setDebug(debug);
+}
+
+/*
+ * Adds to the context structure values to retrieve later it easily by the user
+ */
+
+void Ubidots::addContext(char *key_label, char *key_value) {
+  (_context + _current_context)->key_label = key_label;
+  (_context + _current_context)->key_value = key_value;
+  _current_context++;
+  if (_current_context >= MAX_VALUES) {
+    if (_debug) {
+      Serial.println(F("You are adding more than the maximum of consecutive "
+                       "key-values pairs"));
+    }
+    _current_context = MAX_VALUES;
+  }
+}
+
+/*
+ * Retrieves the actual stored context properly formatted
+ */
+
+void Ubidots::getContext(char *context_result) {
+  getContext(context_result, _iotProtocol);
+}
+
+void Ubidots::getContext(char *context_result, IotProtocol iotProtocol) {
+  // TCP context type
+  if (iotProtocol == UBI_TCP || iotProtocol == UBI_UDP) {
+    sprintf(context_result, "");
+    for (uint8_t i = 0; i < _current_context;) {
+      sprintf(context_result, "%s%s=%s", context_result,
+              (_context + i)->key_label, (_context + i)->key_value);
+      i++;
+      if (i < _current_context) {
+        sprintf(context_result, "%s$", context_result);
+      } else {
+        sprintf(context_result, "%s", context_result);
+        _current_context = 0;
+      }
+    }
+  }
+
+  // HTTP context type
+  if (iotProtocol == UBI_HTTP) {
+    sprintf(context_result, "");
+    for (uint8_t i = 0; i < _current_context;) {
+      sprintf(context_result, "%s\"%s\":\"%s\"", context_result,
+              (_context + i)->key_label, (_context + i)->key_value);
+      i++;
+      if (i < _current_context) {
+        sprintf(context_result, "%s,", context_result);
+      } else {
+        sprintf(context_result, "%s", context_result);
+        _current_context = 0;
+      }
+    }
+  }
+}
+
+bool Ubidots::serverConnected() { return _cloudProtocol->serverConnected(); }
+
+void Ubidots::getDeviceIMEI(char IMEI[]) {
+  const uint8_t IMEI_SIZE = 20;
+
+  GPRS *gprs = GPRS::getInstance();
+
+  auto getIMEI = [IMEI_SIZE](GPRS *gprs, char *IMEI) -> void {
+    gprs->send("AT+GSN=?");
+
+    char *response = (char *)malloc(IMEI_SIZE * sizeof(char));
+
+    while (true) {
+      int ret = gprs->recv(response, IMEI_SIZE);
+      if (ret <= 0) {
+        break;
+      }
+      response[ret] = '\0';
+    }
+    strcpy(IMEI, response);
+  };
+
+  if (gprs == NULL) {
+    gprs = new GPRS(7, 8, 19200); // RX TX BAUDRATE
```

Comment:
why do we need another client? we should use the static client instance

Reply:
There could be a bug if the user use the getDeviceIMEI before constructing the Ubidots object.

Reply:
It is not possible to call the getDeviceIMEI without instantiating the Ubidots class. In this case, you should get the actual GPRS singleton instance